src/main/tv/codely/scala_intro_examples/lesson_09_oop/exercise_solutions/joan_miralles_solution/Email.scala

```diff
@@ -0,0 +1,23 @@
+package tv.codely.scala_intro_examples.lesson_09_oop.exercise_solutions.joan_miralles_solution
+
+import java.security.InvalidParameterException
+
+object Email {
+  def apply(completeEmail: String): Email = {
+    val atIndex = completeEmail.indexOf("@")
+    if (completeEmail.isEmpty || !completeEmail.contains("@") || !completeEmail.substring(completeEmail.indexOf("@")).contains(".")) {
+      throw new InvalidParameterException(s"$completeEmail is not a valid e-mail")
+    }
+    Email(completeEmail.substring(0, atIndex), completeEmail.substring(atIndex + 1, completeEmail.length))
```

Comment:
Suggestion: Extract to inner case classes in order to provide more semantics and move the domain logic to their corresponding domain models:

From:
```scala
Email(completeEmail.substring(0, atIndex), completeEmail.substring(atIndex + 1, completeEmail.length))
```

To:
```scala
Email(Local.fromEmail(completeEmail), Domain.fromEmail(completeEmail))

object Local {
  def fromEmail(email: String): Local = email.substring(0, email.indexOf("@"))
}

final case class Local(value: String)

object Domain {
  // … you get it 🙂
}

final case class Domain(value: String)
```